EDUH-2.2

DimaSzZ opened the pull request:
infrastructure & Migrator

MrazotaIzTirasa committed:
EDUH-2.2

MrazotaIzTirasa committed:
EDUH 2.2
Разработка
Прописал конфиг и DbContext

MrazotaIzTirasa committed:
EDUH-2.2
Создал 3 мигратор, реализацию репозиториев и прописал конфиги для проекта Data

MrazotaIzTirasa committed:
EDUH-2.2
Разработка + правка багов
Поправил реализацию UoW и косметические правки

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/CourseConfiguration.cs:
```diff
@@ -0,0 +1,36 @@
+        
+        builder.Property(c => c.EducatorId)
+            .HasColumnName("educator_id")
+            .IsRequired();
```
Можно убрать здесь IsRequired, так как ниже уже прописано для FK IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,60 @@
+            fullName.Property(f => f.FirstName)
+                .IsRequired()
+                .HasMaxLength(60)
+                .HasAnnotation("MinLength", 2)
```
А это будет работать ?

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,60 @@
+        
+        builder.Property(e => e.Gender)
+            .HasColumnName("gender")
+            .IsRequired();
```
Еще не хватает указать значение по умолчанию

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,60 @@
+        {
+            phone.Property(p => p.Value)
+                .HasColumnName("phone")
+                .IsRequired();
```
Ограничь в бд номер тела хотя бы по длине строки

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EnrollmentConfiguration.cs:
```diff
@@ -0,0 +1,36 @@
+        
+        builder.HasOne<Student>()
+            .WithMany(s => s.Enrollments)
+            .HasForeignKey(e => e.StudentId);
```
Этот ключ должен быть обязательным

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EnrollmentConfiguration.cs:
```diff
@@ -0,0 +1,36 @@
+        
+        builder.HasOne<Course>()
+            .WithOne()
+            .HasForeignKey<Enrollment>(e => e.CourseId);
```
Этот ключ должен быть обязательным

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,82 @@
+            
+            builder.Property(s => s.Gender)
+                .HasColumnName("gender")
+                .IsRequired();
```
Еще надо значение по умолчанию

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,82 @@
+            {
+                phone.Property(p => p.Value)
+                    .HasColumnName("phone")
+                    .IsRequired();
```
Телефон у каждого уникален, нужен индекс

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,82 @@
+                .Property(e => e.Value)
+                .HasColumnName("email")
+                .HasMaxLength(255)
+                .IsRequired();
```
Эл. почта у каждого уникальна, нужен индекс

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,82 @@
+                    .HasColumnName("city")
+                    .IsRequired();
+                
+                address.Property(a => a.Street)
```
Не вижу ограничение по длине строки для Street и City

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EnrollmentConfiguration.cs:
```diff
@@ -0,0 +1,36 @@
+            .HasColumnName("course_id");
+        
+        builder.Property(e => e.EnrollmentDate)
+            .HasColumnName("enrollment_date");
```
Для дат лучше указывать явно тип для БД posgreSql, есть свои приколы. Погляди в инетике

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/DbContext/AppDbContext.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <param name="builder"></param>
+    protected override void OnModelCreating(ModelBuilder builder)
+    {
+        base.OnModelCreating(builder);
```
Этот метод вызывать не нужно

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/DbContext/AppDbContext.cs:
```diff
@@ -0,0 +1,35 @@
+using EduHub.StudentService.Infrastructure.Data.Configuration;
+using Microsoft.EntityFrameworkCore;
+
+namespace EduHub.StudentService.Infrastructure.Data;
```
namespace надо поправить

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/DbContext/AppDbContext.cs:
```diff
@@ -0,0 +1,35 @@
+        builder.HasPostgresEnum<Gender>();
+    }
+    
+    public DbSet<Student> Students { get; set; }
```
DbSet располагаем после конструктора и до методов

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Migrator/DesignTimeDbContextFactory.cs:
```diff
@@ -0,0 +1,30 @@
+        var connectionString = configuration.GetConnectionString("DefaultConnection");
+        
+        builder.UseNpgsql(connectionString,
+            options => options.MigrationsAssembly("EduHub.StudentService.Infrastructure.Migrator"));
```
Assembly можно вытащить из кода и убрать хард код

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/BaseRepository.cs:
```diff
@@ -0,0 +1,60 @@
+    {
+        var entity = await _context.Set<TEntity>().FindAsync(new object[] {id}, cancellationToken);
+        
+        Guard.Against.Null(entity);
```
Здесь не надо проверять на null

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/CourseRepository.cs:
```diff
@@ -0,0 +1,12 @@
+
+public class CourseRepository : BaseRepository<Course>, ICourseRepository
+{
+    public CourseRepository(DbContext context) : base(context)
```
В спец репах в конструкторе лучше передавать конкретный DbContext, То есть твой AppDbContext

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/UoW/StudentUnitOfWork.cs:
```diff
@@ -0,0 +1,24 @@
+        public StudentUnitOfWork(AppDbContext context, IServiceProvider serviceProvider)
+            : base(context)
+        {
+            _serviceProvider = Guard.Against.Null(serviceProvider);
```
Это лучше убрать в базовую реализацию

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/UoW/UnitOfWork.cs:
```diff
@@ -0,0 +1,50 @@
+    
+    public UnitOfWork(TContext context)
+    {
+        _context = Guard.Against.Null(context);
```
_serviceProvider должен быть здесь, из него и сможешь получить контекст.

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/UoW/UnitOfWork.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <returns>Нужный нам репозиторий</returns>
+    public TRepository GetRepository<TRepository>()
+    {
+        return (TRepository) _context.GetService(typeof(TRepository));
```
Тут лучше написать _serviceProvider..GetRequiredService<TRepository>(); И этот метод GetRepository использовать в StudentUnitOfWork для получения репозиториев

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/UoW/UnitOfWork.cs:
```diff
@@ -0,0 +1,50 @@
+        return (TRepository) _context.GetService(typeof(TRepository));
+    }
+    
+    public TContext Context => _context;
```
публик контекст тут вроде и не нужен

MrazotaIzTirasa committed:
EDUH.2-2
Исправление логики
Исправил конфиги, подправил репозитории и контекст. Добавил миграции

MrazotaIzTirasa committed:
EDUH-2.2
исправление
хард код

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+        builder.OwnsOne(e => e.Phone, phone =>
+        {
+            phone.Property(p => p.Value)
+                .HasMaxLength(8)
```
Уверен, что 8? 37377812345 это не 8 цифр, а 11

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+                .HasColumnName("phone");
+            
+            phone.HasIndex(p => p.Value)
+                .IsUnique();
```
Для индексов пишем свои имена и лучше их хранить где-нибудь в константе, потом они еще пригодятся

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/CourseConfiguration.cs:
```diff
@@ -0,0 +1,33 @@
+        
+        builder.Property(c => c.Name)
+            .HasColumnName("name")
+            .HasMaxLength(50);
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/CourseConfiguration.cs:
```diff
@@ -0,0 +1,33 @@
+        
+        builder.HasOne<Educator>()
+            .WithMany(e => e.Courses)
+            .HasForeignKey(c => c.EducatorId);
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            fullName.Property(f => f.FirstName)
+                .HasMaxLength(60)
+                .HasColumnName("first_name");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+            
+            fullName.Property(f => f.Surname)
+                .HasMaxLength(60)
+                .HasColumnName("surname");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+                .HasMaxLength(60)
+                .HasColumnName("surname");
+            
+            fullName.Property(f => f.Patronymic)
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+        
+        builder.Property(e => e.Gender)
+            .HasDefaultValue(Gender.Default)
+            .HasColumnName("gender");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            phone.Property(p => p.Value)
+                .HasMaxLength(8)
+                .HasColumnName("phone");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+        });
+        
+        builder.Property(e => e.YearsExperience)
+            .HasColumnName("years_experience");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EducatorConfiguration.cs:
```diff
@@ -0,0 +1,57 @@
+        
+        builder.Property(e => e.DateEmployment)
+            .HasColumnName("date_employment")
+            .HasColumnType("date");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/EnrollmentConfiguration.cs:
```diff
@@ -0,0 +1,39 @@
+        
+        builder.Property(e => e.EnrollmentDate)
+            .HasColumnName("enrollment_date")
+            .HasColumnType("date");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                fullName.Property(f => f.FirstName)
+                    .HasMaxLength(60)
+                    .HasColumnName("first_name");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+                
+                fullName.Property(f => f.Surname)
+                    .HasMaxLength(60)
+                    .HasColumnName("surname");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+                
+                fullName.Property(f => f.Patronymic)
+                    .HasMaxLength(60)
+                    .HasColumnName("patronymic");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+            builder.OwnsOne(s => s.Phone, phone =>
+            {
+                phone.Property(p => p.Value)
+                    .HasMaxLength(8)
```
Также поправить валидацию на макс длину и не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+                    .HasColumnName("phone");
+                
+                phone.HasIndex(p => p.Value)
+                    .IsUnique();
```
Также надо своё имя индекса

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+            });
+            
+            builder.Property(s => s.Avatar)
+                .HasColumnName("avatar");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                email.Property(e => e.Value)
+                    .HasColumnName("email")
+                    .HasMaxLength(255);
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+                    .HasMaxLength(255);
+                
+                email.HasIndex(e => e.Value)
+                    .IsUnique();
```
Также надо своё имя индекса

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                address.Property(a => a.City)
+                    .HasColumnName("city")
+                    .HasMaxLength(100);
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+                
+                address.Property(a => a.Street)
+                    .HasColumnName("street")
+                    .HasMaxLength(100);
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+                    .HasMaxLength(100);
+                
+                address.Property(a => a.NumberHouse)
+                    .HasColumnName("number_house");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Configuration/StudentConfiguration.cs:
```diff
@@ -0,0 +1,79 @@
+            
+            builder.Property(s => s.DateBirth)
+                .HasColumnName("date_birth")
+                .HasColumnType("date");
```
не хватает IsRequired

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/DbContext/AppDbContext.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+    }
+    
+    public DbSet<Student> Students { get; set; }
```
Хорошо ли , что DbSet'ы имеют открытый set ?

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/UoW/UnitOfWork.cs:
```diff
@@ -0,0 +1,53 @@
+    /// <returns>Нужный нам репозиторий</returns>
+    public TRepository GetRepository<TRepository>()
+    {
+        var contextFromServiceProvider = _serviceProvider.GetRequiredService<TRepository>();
```
Да можно и просто return прописать без создания переменной

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Repositories/UoW/UnitOfWork.cs:
```diff
@@ -0,0 +1,53 @@
+        return contextFromServiceProvider;
+    }
+    
+    protected TContext Context => _context;
```
Убери это, нигде же не используется

MrazotaIzTirasa committed:
EDUH-2.2
Рефакторинг кода
Сделал косметические правки, поправил строку подключения в миграторе, исправил конфиги, добавил статические индексы

1nsanov commented on src/Services/StudentService/Infrastructure/EduHub.StudentService.Infrastructure.Data/Constant/IndexNames.cs:
```diff
@@ -0,0 +1,7 @@
+
+public static class IndexNames
+{
+    public static string Phone = "IX_Unique_Phone";
```
Лучше для такого использовать не статик поля, а const

1nsanov reviewed (changes requested):
1. Один коммент в PR
2. Поправить мапинг энама
3. Перегенерить миграцию

MrazotaIzTirasa committed:
EDUH-2.2
Оставшиеся 3 коммента

MrazotaIzTirasa committed:
EDUH-2.2
Миграции йобаные
